asustor.c

```diff
@@ -357,27 +403,165 @@ static struct gpio_chip *find_chip_by_name(const char *name)
+			} else {
+				if (strstr(dmi_val, dmi->matches[i].substr))
+					continue;
+			}
+		}
+
+		/* No match */
+		return false;
+	}
+	return true;
+}
+
+// how often does the PCI(e) device with given vendor/device IDs exist in this system?
+static int count_pci_device_instances(unsigned int vendor, unsigned int device)
+{
+	// start with -1 as the do-while loop runs at least once even if nothing is found
+	int ret            = -1;
+	struct pci_dev *pd = NULL;
+	do {
+		++ret;
+		pd = pci_get_device(vendor, device, pd);
+	} while (pd != NULL);
+	return ret;
+}
+
+// check if sys->pci_matches[] match with the PCI devices in the system
+// NOTE: this only checks if the devices in sys->pci_matches[] exist in the system
+//       with the expected count (or do *not* exist if the counts are 0),
+//       PCI devices existing that aren't listed in sys->pci_matches[] is expected
+//       and does not make this function fail.
+static bool pci_devices_match(const struct asustor_driver_data *sys)
+{
+	int i;
+	for (i = 0; i < ARRAY_SIZE(sys->pci_matches); ++i) {
+		int dev_cnt;
+		const struct pci_device_match *pdm;
+		pdm = &sys->pci_matches[i];
+		if (pdm->vendorID == 0 && pdm->deviceID == 0) {
+			// no more entries, the previous ones matched
+			// or we would've returned false already
+			return true;
+		}
+		dev_cnt = count_pci_device_instances(pdm->vendorID,
+		                                     pdm->deviceID);
+		if (dev_cnt < pdm->min_count || dev_cnt > pdm->max_count) {
+			return false;
+		}
+	}
+	return true;
+}
+
+// find out which ASUSTOR system this is, based on asustor_systems[], including
+// their linked asustor_driver_data's pci_matches
+// returns NULL if this isn't a known system
+static const struct dmi_system_id *find_matching_asustor_system(void)
+{
+	int as_idx;
+
+	for (as_idx = 0; as_idx < ARRAY_SIZE(asustor_systems); as_idx++) {
+		struct asustor_driver_data *dd;
+		const struct dmi_system_id *sys;
+		sys = &asustor_systems[as_idx];
+		dd  = sys->driver_data;
+		if (dd == NULL) {
+			// no driverdata? must be at end of table
+			break;
+		}
+
+		if (!dmi_matches(sys)) {
+			continue;
+		}
+
+		if (!pci_devices_match(dd)) {
+			continue;
+		}
+
+		// DMI and PCI devices matched => this is it
+		return sys;
+	}
+
+	return NULL;
+}
+
+static char *force_device = "";
+module_param(force_device, charp, S_IRUSR | S_IRGRP | S_IROTH);
+MODULE_PARM_DESC(
+	force_device,
+	"Don't try to detect ASUSTOR device, use the given one instead. "
+	"Valid values: " VALID_OVERRIDE_NAMES);
+
 // TODO(mafredri): Allow force model for testing.
 static int __init asustor_init(void)
 {
 	const struct dmi_system_id *system;
 	const struct gpiod_lookup *keys_table;
 	int ret, i;
 
-	system = dmi_first_match(asustor_systems);
-	if (!system) {
-		pr_info("No supported ASUSTOR mainboard found");
-		return -ENODEV;
-	}
+	driver_data = NULL;
+	// allow overriding detection with force_device kernel parameter
+	if (force_device && *force_device) {
```

Comment:
```suggestion
	if (*force_device) {
```
Is the non-NULL check actually necessary? I see we already initialized it to the empty string.

Reply:
not sure TBH, I figured it doesn't hurt